src/Components/WebAssembly/Server/src/ComponentsWebAssemblyApplicationBuilderExtensions.cs

```diff
@@ -16,74 +15,50 @@ namespace Microsoft.AspNetCore.Builder
     /// <summary>
     /// Extensions for mapping Blazor WebAssembly applications.
     /// </summary>
-    public static class ComponentsWebAssemblyEndpointRouteBuilderExtensions
+    public static class ComponentsWebAssemblyApplicationBuilderExtensions
     {
         /// <summary>
         /// Maps a Blazor webassembly application to the <paramref name="pathPrefix"/>.
         /// </summary>
-        /// <param name="endpoints">The <see cref="IEndpointRouteBuilder"/>.</param>
+        /// <param name="builder">The <see cref="IApplicationBuilder"/>.</param>
         /// <param name="pathPrefix">The <see cref="PathString"/> that indicates the prefix for the Blazor application.</param>
-        /// <returns>The <see cref="IEndpointConventionBuilder"/></returns>
-        public static IEndpointConventionBuilder MapBlazorWebAssemblyApplication(this IEndpointRouteBuilder endpoints, PathString pathPrefix)
+        /// <returns>The <see cref="IApplicationBuilder"/></returns>
+        public static IApplicationBuilder MapBlazorFrameworkFiles(this IApplicationBuilder builder, PathString pathPrefix)
         {
-            if (endpoints is null)
+            if (builder is null)
             {
-                throw new ArgumentNullException(nameof(endpoints));
+                throw new ArgumentNullException(nameof(builder));
             }
 
-            var webHostEnvironment = endpoints.ServiceProvider.GetRequiredService<IWebHostEnvironment>();
+            var webHostEnvironment = builder.ApplicationServices.GetRequiredService<IWebHostEnvironment>();
 
             var options = CreateStaticFilesOptions(webHostEnvironment.WebRootFileProvider);
-            var appBuilder = endpoints.CreateApplicationBuilder();
 
-            appBuilder.Use(async (ctx, next) =>
+            builder.MapWhen(ctx => ctx.Request.Path.StartsWithSegments(pathPrefix, out var rest) && rest.StartsWithSegments("/_framework") && !rest.StartsWithSegments("/_framework/blazor.server.js"),
+            subBuilder =>
             {
-                var endpoint = ctx.GetEndpoint();
-                try
+                subBuilder.Use(async (ctx, next) =>
```

Comment:
Out of interest, would an alternative working implementation have been to configure the "on prepare response" callback for `options` and put the header controls in there, then just have `subBuilder.UseStaticFiles(options)` without the extra bit of middleware?

I'm not saying that's better. Just want to understand the options.

Reply:
I just did the most straightforward transformation from what we had before, using OnPrepareResponse would have also worked.

Reply:
I think the difference you'd seen in behavior here is that OnPrepareResponse would handle the cases where the file was found - this will add a header for a 404.

Reply:
Yeah, I think you are right, not a big deal since I plan to work on this for preview 3 and a 404 can’t be cached anyways?